front/src/applications/operationalStudies/utils.ts

```diff
@@ -293,3 +296,93 @@ export const isTooFast = (
   }
   return false;
 };
+
+/**
+ * Adjusts a step's arrival time based on its own tolerance values and the given direction.
+ *
+ * @param payload - The step containing timing data.
+ * @param direction - 'upstream' or 'downstream'
+ * @returns The updated step with adjusted arrival time and tolerances.
+ */
+export const adjustPayloadByDirection = (
```

Comment:
I think you can make this function easier to read:
- if the map, early return if the step is not a extremity (`index !== 0 || index !== payload.body.steps.length`) or if it is an extremity and does not have a timing_data (if it has a timing_data, then its arrival_time is defined)
- add blank lines to make the code easier to read
- `step.timing_data.arrival_time_tolerance_before` and `step.timing_data.arrival_time_tolerance_after` is very long to read, destructure it to simplify the code (you can rename it `beforeTolerance` and `afterTolerance` for ex)
- try to rationalize the number of tests: here you test three times `direction === 'downstream'`, but you can do it only once